blocks/bricks/__init__.py

```diff
@@ -1,22 +1,324 @@
+        >>> class Foo(Brick):
+        ...     @application(outputs=['baz'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.property('inputs')
+        ...     def apply_inputs(self):
+        ...         return ['foo', 'bar']
+        >>> class Bar(Brick):
+        ...     def __init__(self, foo):
+        ...         self.foo = foo
+        ...
+        ...     @application(outputs=['foo'])
+        ...     def apply(self, x):
+        ...         return x + 1
+        ...
+        ...     @apply.delegate
+        ...     def apply_delegate(self):
+        ...         return self.foo.apply
+        >>> foo = Foo()
+        >>> bar = Bar(foo)
+        >>> bar.apply.outputs
+        ['foo']
+        >>> bar.apply.inputs
+        ['foo', 'bar']
+
+        """
+        self.delegate_function = f
+        return f
+
+    def __get__(self, instance, owner):
+        """Instantiate :class:`BoundedApplication` for each :class:`Brick`."""
+        if instance is None:
+            return self
+        elif instance in self.instances:
+            return self.instances[instance]
+        else:
+            bounded_application = BoundApplication(self, instance)
+            self.instances[instance] = bounded_application
+            return bounded_application
+
+    def __getattr__(self, name):
+        # Mimic behaviour of properties
+        if 'properties' in self.__dict__ and name in self.properties:
+            return property(create_unbound_method(self.properties[name],
+                                                  self.brick))
+        raise AttributeError
+
+    def __setattr__(self, name, value):
+        # Mimic behaviour of read-only properties
+        if 'properties' in self.__dict__ and name in self.properties:
+            raise AttributeError("can't set attribute")
+        super(Application, self).__setattr__(name, value)
+
+    @property_
+    def inputs(self):
+        return self._inputs
+
+    @inputs.setter
+    def inputs(self, inputs):
+        args_names, varargs_name, _, _ = inspect.getargspec(
+            self.application)
+        if not all(input_ in args_names + [varargs_name] for input_ in inputs):
+            raise ValueError("Unexpected inputs")
+        self._inputs = inputs
+
+    @property_
+    def name(self):
+        return self.application.__name__
+
+    def __call__(self, application, *args, **kwargs):
+        return_dict = kwargs.pop('return_dict', False)
+        return_list = kwargs.pop('return_list', False)
+        if return_list and return_dict:
+            raise ValueError
+
+        brick = application.brick
+
+        # Find the names of the inputs to the application method
+        args_names, varargs_name, _, _ = inspect.getargspec(
+            self.application)
+        args_names = args_names[1:]
+
+        # Construct the ApplicationCall, used to store data in for this call
+        call = ApplicationCall(brick, application)
```

Comment:
We should update the documentation of the `ApplicationCall` class to reflect the fact that its `application` attribute is now a `BoundApplication` instance.